test_rclcpp/test/test_multithreaded.cpp

```diff
@@ -308,40 +338,72 @@ static inline void multi_access_publisher(bool intra_process)
     sub_callback,
     subscription_options);
 
-  // wait a moment for everything to initialize
+  // wait a moment for everything to initialize (pub/subs matching)
   test_rclcpp::wait_for_subscriber(node, node_topic_name);
 
-  // use atomic
-  std::atomic_uint timer_counter(0);
-
+  // timers
+  std::atomic_uint publish_counter(0);
   auto timer_callback =
-    [&executor, &pub, &timer_counter, &subscription_counter, &num_messages](
+    [&executor, &pub, &publish_counter, &subscription_counter, &num_messages](
     rclcpp::TimerBase & timer)
     {
       auto msg = std::make_unique<test_rclcpp::msg::UInt32>();
-      if (timer_counter.load() >= num_messages) {
+      auto next_publish_count = ++publish_counter;
+      if (next_publish_count > num_messages) {
+        // enough messages have been sent, step the counter back, and cancel timer
+        publish_counter--;
```

Comment:
Hm, this bit feels non-atomic to me.  I understand that this should only happen after we've published enough messages, and I honestly can't come up with a scenario where this fails.  But do we really need to do this decrement?  Could we instead just let it increment too high unnecessarily, then use `num_messages` when doing checks below?

Reply:
> Could we instead just let it increment too high unnecessarily, then use num_messages when doing checks below?

Actually, sorry, this is dumb.  This won't work because we are comparing `num_messages` to `publish_counter`.

That said, I wonder if we can just drop the `publish_counter` checks below, and instead just rely on the `subscription_counter`.  At the end of the day, that's the thing we are most interested in.  Thoughts?

Reply:
Well, I could do that, though it's nice to see that we actually did publish the right number of times, and that accessing the counter for that publish is thread-safe (needing the atomics and this back and forth business). We could use a lock to check the value before incrementing it (avoiding the roll back) but I was trying to keep with the atomics.

I'd like to merge this incremental improvement and open a new pr for using a lock (should be clearer to read), but I want this to go in before the freeze ideally.

Reply:
Ok, I think I resolved your concern in [b7c988e](https://github.com/ros2/system_tests/pull/515/commits/b7c988ee8ad7a057a52c9c7d9bf2938a8812327d), at least it's clearer to me. Though I'm still checking all three "counters".